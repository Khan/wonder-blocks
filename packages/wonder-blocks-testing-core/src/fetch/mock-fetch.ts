import {fetchRequestMatchesMock} from "./fetch-request-matches-mock";
import {mockRequester} from "../mock-requester";
import type {FetchMockFn, FetchMockOperation} from "./types";

/**
 * A mock for the fetch function passed to GqlRouter.
 */
export const mockFetch = (): FetchMockFn =>
    mockRequester<FetchMockOperation>(
        fetchRequestMatchesMock,
        // NOTE(somewhatabstract): The indentation is expected on the lines
        // here.
        (input, init) =>
            `Input: ${
                typeof input === "string"
                    ? input
                    : JSON.stringify(input, null, 2)
            }
    Options: ${init == null ? "None" : JSON.stringify(init, null, 2)}`,
    );
