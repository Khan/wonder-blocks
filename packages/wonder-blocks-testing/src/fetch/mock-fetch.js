// @flow
import {fetchRequestMatchesMock} from "./fetch-request-matches-mock";
import {mockRequester} from "../mock-requester";
import type {FetchMockFn, FetchMockOperation} from "./types";

/**
 * A mock for the fetch function passed to GqlRouter.
 */
export const mockFetch = (): FetchMockFn =>
    mockRequester<FetchMockOperation, _>(
        fetchRequestMatchesMock,
        (input, init) =>
            `Input: ${
                typeof input === "string"
                    ? input
                    : JSON.stringify(input, null, 2)
            }
    Options: ${init == null ? "None" : JSON.stringify(init, null, 2)}`,
    );
