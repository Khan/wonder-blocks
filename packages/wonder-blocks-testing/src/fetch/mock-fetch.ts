import {fetchRequestMatchesMock} from "./fetch-request-matches-mock";
import {mockRequester} from "../mock-requester";
import type {FetchMockFn, FetchMockOperation} from "./types";

/**
 * A mock for the fetch function passed to GqlRouter.
 */
// @ts-expect-error [FEI-5019] - TS2709 - Cannot use namespace '_' as a type.
export const mockFetch = (): FetchMockFn =>
    mockRequester<FetchMockOperation, _>(
        fetchRequestMatchesMock,
        (input, init) =>
            `Input: ${
                typeof input === "string"
                    ? input
                    : JSON.stringify(input, null, 2)
            }
Options: ${init == null ? "None" : JSON.stringify(init, null, 2)}`,
    );
