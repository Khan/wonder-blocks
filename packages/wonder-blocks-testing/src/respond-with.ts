import {SettleSignal} from "./settle-signal";
import {ResponseImpl} from "./response-impl";
import type {GraphQLJson} from "./types";

// We want the parameterization here so that folks can assert a response is
// of a specific type if passing between various functions. For example,
// the graphql mocking framework might want to assert a response is returning
// the expected data structure. We could use `opaque` but that would then
// hide the `toPromise` call we want to provide.
/* eslint-disable no-unused-vars */
/**
 * Describes a mock response to a fetch request.
 */
export type MockResponse<TData> = {
    /**
     * Create a promise from the mocked response.
     *
     * If a signal was provided when the mock response was created, the promise
     * will only settle to resolution or rejection if the signal is raised.
     */
    readonly toPromise: () => Promise<Response>;
};
/* eslint-enable no-unused-vars */

type InternalMockResponse =
    | {
          readonly type: "text";
          readonly text: string | (() => string);
          readonly statusCode: number;
          readonly signal: SettleSignal | null | undefined;
      }
    | {
          readonly type: "reject";
          readonly error: Error | (() => Error);
          readonly signal: SettleSignal | null | undefined;
      };

/**
 * Helper for creating a text-based mock response.
 */
const textResponse = <TData>(
    text: string | (() => string),
    statusCode: number,
    signal?: SettleSignal | null,
): MockResponse<TData> => ({
    toPromise: () =>
        makeMockResponse({
            type: "text",
            text,
            statusCode,
            signal,
        }),
});

/**
 * Helper for creating a rejected mock response.
 */
const rejectResponse = (
    error: Error | (() => Error),
    signal?: SettleSignal | null,
): MockResponse<never> => ({
    toPromise: () =>
        makeMockResponse({
            type: "reject",
            error,
            signal,
        }),
});

/**
 * Helpers to define mock responses for mocked requests.
 */
export const RespondWith = Object.freeze({
    /**
     * Response with text body and status code.
     * Status code defaults to 200.
     */
    text: <TData = string>(
        text: string,
        statusCode: number = 200,
        signal: SettleSignal | null = null,
    ): MockResponse<TData> => textResponse<TData>(text, statusCode, signal),

    /**
     * Response with JSON body and status code 200.
     */
    json: <TJson extends Record<any, any>>(
        json: TJson,
        signal: SettleSignal | null = null,
    ): MockResponse<TJson> =>
        textResponse<TJson>(() => JSON.stringify(json), 200, signal),

    /**
     * Response with GraphQL data JSON body and status code 200.
     */
    graphQLData: <TData extends Record<any, any>>(
        data: TData,
        signal: SettleSignal | null = null,
    ): MockResponse<GraphQLJson<TData>> =>
        textResponse<GraphQLJson<TData>>(
            () => JSON.stringify({data}),
            200,
            signal,
        ),

    /**
     * Response with body that will not parse as JSON and status code 200.
     */
    unparseableBody: (signal: SettleSignal | null = null): MockResponse<any> =>
        textResponse("INVALID JSON", 200, signal),

    /**
     * Rejects with an AbortError to simulate an aborted request.
     */
    abortedRequest: (signal: SettleSignal | null = null): MockResponse<any> =>
        rejectResponse(() => {
            const abortError = new Error("Mock request aborted");
            abortError.name = "AbortError";
            return abortError;
        }, signal),

    /**
     * Rejects with the given error.
     */
    reject: (
        error: Error,
        signal: SettleSignal | null = null,
    ): MockResponse<any> => rejectResponse(error, signal),

    /**
     * A non-200 status code with empty text body.
     * Equivalent to calling `ResponseWith.text("", statusCode)`.
     */
    errorStatusCode: (
        statusCode: number,
        signal: SettleSignal | null = null,
    ): MockResponse<any> => {
        if (statusCode < 300) {
            throw new Error(`${statusCode} is not a valid error status code`);
        }
        return textResponse("{}", statusCode, signal);
    },

    /**
     * Response body that is valid JSON but not a valid GraphQL response.
     */
    nonGraphQLBody: (signal: SettleSignal | null = null): MockResponse<any> =>
        textResponse(
            () =>
                JSON.stringify({
                    valid: "json",
                    that: "is not a valid graphql response",
                }),
            200,
            signal,
        ),

    /**
     * Response that is a GraphQL errors response with status code 200.
     */
    graphQLErrors: (
        errorMessages: ReadonlyArray<string>,
        signal: SettleSignal | null = null,
    ): MockResponse<GraphQLJson<any>> =>
        textResponse<GraphQLJson<any>>(
            () =>
                JSON.stringify({
                    errors: errorMessages.map((e) => ({
                        message: e,
                    })),
                }),
            200,
            signal,
        ),
});

const callOnSettled = (
    signal: SettleSignal | null | undefined,
    fn: () => void,
): void => {
    if (signal == null || signal.settled) {
        fn();
        return;
    }

    const onSettled = () => {
        signal.removeEventListener("settled", onSettled);
        fn();
    };
    signal.addEventListener("settled", onSettled);
};

/**
 * Turns a MockResponse value to an actual Response that represents the mock.
 */
const makeMockResponse = (
    response: InternalMockResponse,
): Promise<Response> => {
    const {signal} = response;

    switch (response.type) {
        case "text":
            return new Promise((resolve, reject: (error?: any) => void) => {
                callOnSettled(signal, () => {
                    const text =
                        typeof response.text === "function"
                            ? response.text()
                            : response.text;
                    resolve(
                        new ResponseImpl(text, {status: response.statusCode}),
                    );
                });
            });

        case "reject":
            return new Promise((resolve, reject: (error?: any) => void) => {
                callOnSettled(signal, () =>
                    reject(
                        response.error instanceof Error
                            ? response.error
                            : response.error(),
                    ),
                );
            });

        /* istanbul ignore next */
        default:
            if (process.env.NODE_ENV !== "production") {
                // If we're not in production, give an immediate signal that the
                // dev forgot to support this new type.
                // @ts-expect-error [FEI-5019] - TS2339 - Property 'type' does not exist on type 'never'.
                throw new Error(`Unknown response type: ${response.type}`);
            }
            // Production; assume a rejection.
            return makeMockResponse({
                type: "reject",
                error: new Error("Unknown response type"),
                signal,
            });
    }
};
