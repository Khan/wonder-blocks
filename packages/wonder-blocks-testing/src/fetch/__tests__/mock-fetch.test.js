// @flow
import {Request} from "node-fetch";
import {RespondWith} from "../../respond-with.js";
import {mockFetch} from "../mock-fetch.js";

describe("#mockFetch", () => {
    it.each`
        input                                    | init
        ${"http://example.com/foo"}              | ${undefined}
        ${new URL("http://example.com/foo")}     | ${{method: "GET"}}
        ${new Request("http://example.com/foo")} | ${{method: "POST"}}
    `(
        "should reject with a useful error when there are no matching mocks for %s",
        async ({input, init}) => {
            // Arrange
            const mockFn = mockFetch();

            // Act
            const result = mockFn(input, init);

            // Assert
            await expect(result).rejects.toThrowErrorMatchingSnapshot();
        },
    );

    describe("mockOperation", () => {
        it("should match a similar operation", async () => {
            // Arrange
            const mockFn = mockFetch();
            const operation = "http://example.com/foo";

            // Act
            mockFn.mockOperation(operation, RespondWith.text("TADA!"));
            const result = mockFn(operation, {method: "GET"});

            // Assert
            await expect(result).resolves.toBeDefined();
        });

        it("should not match a different operation", async () => {
            // Arrange
            const mockFn = mockFetch();
            const operation = "http://example.com/foo";

            // Act
            mockFn.mockOperation(operation, RespondWith.text("TADA!"));
            const result = mockFn("http://example.com/bar", {method: "GET"});

            // Assert
            await expect(result).rejects.toThrowError();
        });
    });

    describe("mockOperationOnce", () => {
        it("should match once", async () => {
            // Arrange
            const mockFn = mockFetch();
            const operation = "http://example.com/foo";

            // Act
            mockFn.mockOperationOnce(operation, RespondWith.text("TADA!"));
            const result = mockFn(operation, {method: "GET"});

            // Assert
            await expect(result).resolves.toBeDefined();
        });

        it("should only match once", async () => {
            // Arrange
            const mockFn = mockFetch();
            const operation = "http://example.com/foo";

            // Act
            mockFn.mockOperationOnce(operation, RespondWith.text("TADA!"));
            const result = Promise.all([
                mockFn(operation, {method: "GET"}),
                mockFn(operation, {method: "POST"}),
            ]);

            // Assert
            await expect(result).rejects.toThrowError();
        });
    });
});
