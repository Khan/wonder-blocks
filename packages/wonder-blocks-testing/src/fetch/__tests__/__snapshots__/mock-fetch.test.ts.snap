// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#mockFetch should reject with a useful error when there are no matching mocks for %s 1`] = `
"No matching mock response found for request:
    Input: http://example.com/foo
Options: None"
`;

exports[`#mockFetch should reject with a useful error when there are no matching mocks for %s 2`] = `
"No matching mock response found for request:
    Input: "http://example.com/foo"
Options: {
  "method": "GET"
}"
`;

exports[`#mockFetch should reject with a useful error when there are no matching mocks for %s 3`] = `
"No matching mock response found for request:
    Input: {
  "size": 0,
  "timeout": 0,
  "follow": 20,
  "compress": true,
  "counter": 0
}
Options: {
  "method": "POST"
}"
`;
